File: cvt_aln_fasta.pl
#! c:\perl\bin\perl

use Statistics::Basic qw(:all);

# open directory

my $seq_dir = "RF00029_60-blastclust-sub1-nh";
my $updated;
my $neg_dir = "z_score";
my $err_file = "error";
my $seq_file = "file4.txt";
my $fasta_file;
my $mfe_file;
my @mfe_array;
my $combo;
my $score_file;
my $j;
my $pos_mfe;
my $mfe;
my $pos_rnaali;
my @name;
my $nm;
my @seq;
my $sequence;

my $neg_path = ".\\".$seq_dir."\\".$neg_dir;

# print "$neg_path\n";

mkdir $neg_path;

unless ( open(SEQS, "$seq_dir\\$seq_file") ) {

	print "Cannot open file \"$seq_dir\\$seq_file\"\n\n";
	exit;
}

# Read RNAfold records and store in array

my @allfiles = <SEQS>;

# Close the file

close SEQS;


foreach my $file (@allfiles) {

	# initialize arrays

	@mfe_array = ();
	@seq = ();
	@name = ();
	$sequence = "";

	# find mfe and structure

	chomp($file);

	unless ( open(SEQFILE, "$seq_dir\\$file") ) {

		print "Cannot open file \"$seq_dir\\$file\"\n\n";
		exit;
	}

	# Read Clustal file

	my @pos_rnafold_lines = <SEQFILE>;

	# Close the file

	close SEQFILE;

	my $i = 0;

	foreach my $line (@pos_rnafold_lines) {

		chomp($line);

	# discard blank lines

		if ($line =~ /^$/) {
			$i = 0;
			next;

		}


	# discard header line

		elsif ($line =~ /^CLUSTAL/) {
			next;

		}

	# discard similarity line

		elsif ($line =~ /^\s\*/) {
			next;

		}

	# sequence line

		elsif($line =~ /^[a-z]/) {
			$nm = $line;
			$nm =~ s/\s[-UACG]*$//;
			$name[$i] = $nm;
			$sequence = $line;
			$sequence =~ s/^[a-z].*\s//;
			$seq[$i] .= $sequence;
			$i++;

		}
	}

	$fasta_file = $file;
	$fasta_file =~ s/fin/fasta/;
	$mfe_file = $file;
	$mfe_file =~ s/fin/mfe/;

	unless ( open(FASTA, "> $seq_dir\\$fasta_file") ) {

		print "Cannot open file \"$seq_dir\\$fasta_file\"\n\n";
		exit;
	}



	while ($i > 0) {
		$i--;
		print FASTA ">$name[$i]\n";
		print FASTA "$seq[$i]\n";
	}

	my $outmfe = `rnafold <$seq_dir\\$fasta_file >$seq_dir\\$mfe_file`;

	# Close and delete fasta file

	close FASTA;

	unlink "$seq_dir\\$fasta_file";

	unless ( open(MFE, "$seq_dir\\$mfe_file") ) {

		print "Cannot open file \"$seq_dir\\$mfe_file\"\n\n";
		exit;
	}

	my @mfe_lines = <MFE>;

	# Close and delete MFE file

	close MFE;

	unlink "$seq_dir\\$mfe_file";

	foreach $line (@mfe_lines) {

		if ($line =~ /\(-[0-9]/){
			$mfe = substr($line, -9);
			$mfe =~ s/\)//;
			$mfe =~ s/\(//;
			
		#	print "mfe $mfe\n";

			push @mfe_array, $mfe;
		}

	}

	$combo = $file;
	$combo =~ s/-fin.txt//;

	my $mean = mean(@mfe_array);

	print "$combo,$mean\n";

}

File: cvt_aln_fasta.py
#! /usr/bin/python

###################################################################################################################################################
#
#
#   cvt_aln_fasta.py - convert CLUSTAL alignment to FASTA file format to input to rnafold and calculate mean free energy ot he consensus structure
#
#   Input:      seq_file - list of alignment files to process in CLUSTAL format
#   Output:     mean free energy of the consensus structure output to standard output
#
#
###################################################################################################################################################


import re
from statistics import mean
import os

# patterns

blankline_pattern = re.compile(r'^\w*$')
similarity_line_pattern = re.compile(r'^\s\*')
sequence_line_pattern = re.compile(r'^[a-z]')
mfe_line_pattern = re.compile(r'^[\.\(\)]* ')

# files and directories for data

seq_dir = "RF00004_60_blastclust_sub1"
neg_dir = "z_score"
err_file = "error"
seq_file = "file2a.txt"
root = "/home/azureuser/"
neg_path = root + seq_dir + "/" + neg_dir

# header line

print(",".join(["combo", "mfe"]))

# read and parse CLUSTAL file

for file in open(root + seq_dir + "/" + seq_file):
    
	# initialize arrays

	name = []
	d = {}
	sequence = ""

	# strip file entry of whitespace and nl

	file = file.strip()

	# Read Clustal file and store in dictionary

	i = 0
	for line in open(root + seq_dir + "/" + file):
		line = line.strip()
		if blankline_pattern.search(line) != None:
			continue
		elif line.__contains__("CLUSTAL"):
			continue
		elif similarity_line_pattern.search(line) != None:
			continue
		elif sequence_line_pattern.search(line) != None:
			nm, sequence = line.split()
			nm = nm.strip()
			if nm not in name:
				name.append(nm)
				d[nm] = ""
			d[nm] += sequence

	# Create FASTA and MFE output files

	fasta_file = file.replace('-fin','-fasta')
	mfe_file = file.replace('-fin', '-mfe')
	combo = file.replace('-fin.txt', '')

	# Open FASTA file to write

	fasta_file_handle = open(root + seq_dir + "/" + fasta_file, 'w')

	for key,value in d.items():
		fasta_file_handle.write('>' + key +'\n')
		fasta_file_handle.write(value + '\n')

	# Execute RNAfold to find MFE scores

	cmd = 'RNAfold ' + '< ' + root + seq_dir + "/" + file + ' >' + root + seq_dir + "/" + mfe_file
	os.system(cmd)

	# Parse scores and find the mean

	mfe_list = []

	for line in open(root + seq_dir + "/" + mfe_file):
		if mfe_line_pattern.search(line) != None:
			mfe = line[-9:]
			mfe = mfe.replace('(', '')
			mfe = mfe.replace(')', '')
			mfe = float(mfe.strip())
			mfe_list.append(mfe)
	mfe_mean = mean(mfe_list)

	# Print the combination and mean for each alignment

	print(",".join([combo, str(mfe_mean)]))

	# Remove temporary FASTA and MFE files

	for f1 in [root + seq_dir + "/" + fasta_file, root + seq_dir + "/" + mfe_file]:
		os.remove(f1)